File: LeetCode/Number of Islands.cpp
class Solution {
public:
    int dfs(vector<vector<char>> &grid, int i, int j, set<string> &visited) {
        if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] == '0')
            return 0;
        
        string node = to_string(i) + "," + to_string(j);
        
        if (visited.count(node)) return 0;
        
        visited.insert(node);
        
        dfs(grid, i-1, j, visited);
        dfs(grid, i+1, j, visited);
        dfs(grid, i, j-1, visited);
        dfs(grid, i, j+1, visited);
        
        return 1;
    }
        
    int numIslands(vector<vector<char>>& grid) {
        set<string> visited;
        int cnt = 0;
        
        for(int i=0; i < grid.size(); i++) {
            for(int j=0; j < grid[0].size(); j++) {
                cnt += dfs(grid, i, j, visited);
            }
        }
        
        return cnt;
    }
};

File: LeetCode/Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visited = [[False for _ in range(len(grid[0]))] for __ in range(len(grid))]
        ans = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if visited[i][j] or grid[i][j] == '0':
                    continue
                
                self.dfs(i, j, grid, visited)
                ans += 1
        
        return ans
                
    
    def dfs(self, i: int, j: int, grid: List[List[str]], visited: List[List[bool]]) -> None:
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]):
            return None
        
        if visited[i][j] or grid[i][j] == '0':
            return None
        
        visited[i][j] = True
        
        self.dfs(i+1, j, grid, visited)
        self.dfs(i-1, j, grid, visited)
        self.dfs(i, j+1, grid, visited)
        self.dfs(i, j-1, grid, visited)